File: CCC '96/C++ Solutions/S5 - Maximum Distance.cpp

/*
Author: Simon Yang
Problem: https://dmoj.ca/problem/ccc96s5
Explanation
-------------------------------------------
Given 2 set of numbers, we must find the maxiumum distance between two elements such that the larger
index must contain a value larger or equal to the value of the index. For example if we have a list X
with an index of i and list Y with the index of j, then the maximum distance would be j - i with y[j] 
bigger than or equal to X[i].
*/

//CODE
//-----------------------------------------------
#include <iostream>

// init 2 array to store the 2 list
int l1[1000];
int l2[1000];

int solve(int x) {
	
	int max = 0; // init maximum distance to 0

	for (int i = 0; i < x; i++) { // loop through list 1
		for (int j = x-1; j >= 0; j--) { // loop through list 2
			if (l2[j] >= l1[i]) {  // check constraints
				if (j-i > max) { // check if current differences is greater than greatest 
					max = j-i;  // set differences to greatest
				}
			}
		}
	}return max; // return greatest
}

int main() {
	int N,x; // init number of test cases,and length of list
	std::cin >> N; // get number of test cases
	for (int i = 0; i < N; i++) { // loop through number of test cases
		std::cin >> x; // get length of list
		
		// store values to the list
		for (int j = 0; j < x; j++) { 
			std::cin >> l1[j];
		}
		for (int j = 0; j < x; j++) {
			std::cin >> l2[j];
		}
			
		// print out the maximum distance
		std::cout << "The maximum distance is " << solve(x) << "\n\n";
	}
}

File: CCC '96/Python Solutions/S5 - Maximum Distance.py
"""
Author: Simon Yang
Problem: https://dmoj.ca/problem/ccc96s5
Explanation
-------------------------------------------
Given 2 set of numbers, we must find the maxiumum distance between two elements such that the larger
index must contain a value larger or equal to the value of the index. For example if we have a list X
with an index of i and list Y with the index of j, then the maximum distance would be j - i with y[j] 
bigger than or equal to X[i].
"""

# CODE
#-----------------------------------------------
N = int(input())  # get numbers of test cases

for i in range(N):  # loop through all test cases
	X = int(input()) # get length of list

	# get string and seperate them into int and store them in a list
	l1 = list(map(int,input().split(' ')))
	l2 = list(map(int,input().split(' ')))
	
	maximum = 0  # init maximum
	for j in range(X):  # loop through the list
		for k in reversed(range(X)):  # loop through the list
			if (l2[k] >= l1[j] and k-j>maximum):  # check constraints
				maximum = k-j  # set differences to maximum
	print(f"The maximum distance is {maximum}")  # print out maximum distance